### PR TITLE
Shutdown the context before context's destructor is invoked in tests

### PR DESCRIPTION
**Pull request by ahcorde**

### CODE REVIEW

**Comment by alsora** [comment]
Why this change? The context is initialized in the `SetupTestCase` method so it should be shut down there

**Review by fujitatomoya — APPROVED** [review]
lgtm. the fix itself makes sense to me that we call `shutdown` explicitly once it completes the all test cases with this fixture.

i would also like to know @alsora 's comment above. i think `default_context` will be destructed when the process space is destroyed, then context will be shutdown. What kind of problem did you have with rmw_zenoh?

**Comment by Yadunund** [comment]
@ahcorde any chance you opened this PR because you ran into the `panic` documented here? https://github.com/ros2/rmw_zenoh/issues/170

If so I wonder if my comment [here](https://github.com/ros2/rmw_zenoh/issues/170#issuecomment-2362533300) explains why we need this PR.

**Review by Yadunund — APPROVED** [review]
Please see https://github.com/ros2/rmw_zenoh/issues/170#issuecomment-2364480265 for why this is needed.

TLDR; Leaving the context's destructor to shutdown the context at program termiantion causes Zenoh to panic since Thread local storage gets cleared earlier.

**Comment by Yadunund** [comment]
CI started with [this repos file](https://gist.githubusercontent.com/Yadunund/fc9014a21560fe9728ff03ade81b3009/raw/dad8164d95cbde0d6f04c45601dc87c2e361e1b2/ros2.repos).
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21864/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16134/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1371/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22576/)

**Comment by Yadunund** [comment]
I took the liberty to adopt the change into other tests in `rclcpp` and `rclcpp_lifecyle`. Now we're able to run all the tests with `rmw_zenoh` without anything crashing! 🎉 

@fujitatomoya @alsora kindly review this PR whenever time permits!

Re-run CI after pushing changes:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21865/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16135/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1372/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22577/)

File: rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp [hunk 1]
```diff
@@ -64,6 +64,11 @@ class TestNodeGraph : public ::testing::Test
     rclcpp::shutdown();
   }
 
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
```
Comment (clalancette) [inline rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp:70]: This one doesn't seem right.  Here we are calling `rclcpp::init` in `SetUp`, so either `rclcpp::shutdown` in `TearDown` should be enough, or we should split `rclcpp::init` into `SetUpTestCase`.

**Review by clalancette — CHANGES_REQUESTED** [review]
Generally, these look good to me, but there is one that I think we should revamp.

File: rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp (hunk 1 above)
Comment (ahcorde) [reply rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp]: https://github.com/ros2/rclcpp/pull/2633/commits/9af7a4d149f7b7842a47242097222560bd4ec141

**Comment by ahcorde** [comment]
Pulls: ros2/rclcpp#2633
Gist: https://gist.githubusercontent.com/ahcorde/8d7c14e539372d6cfc1b4315d0a89483/raw/cd27611aeba4eb0fc8ae1722e70b2caaa82ac72d/ros2.repos
BUILD args:  --packages-above-and-dependencies rclcpp rclcpp_lifecycle rclcpp_components --packages-above-and-dependencies rclcpp rclcpp_lifecycle rclcpp_components
TEST args:  --packages-above rclcpp rclcpp_lifecycle rclcpp_components --packages-above rclcpp rclcpp_lifecycle rclcpp_components
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14596
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21876/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16145/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1382/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22590/)

**Comment by Yadunund** [comment]
> Still failing here
> 
> ```
> 44: /home/ahcorde/ros2_rolling/src/ros2/rclcpp/rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp:646: Failure
> 44: Value of: endpoint_gid_is_all_zeros
> 44:   Actual: true
> 44: Expected: false
> ```

This is something we need to fix in `rmw_zenoh`. The changes here are just to get all the tests to run without crashing.

@ahcorde looks like uncrustify is failing.

**Comment by Yadunund** [comment]
Jobs:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21880/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16148/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1385/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22593/)
